dali/python/nvidia/dali/_conditionals.py

```diff
@@ -516,14 +522,34 @@ def if_stmt(self, cond, body, orelse, get_state, set_state, symbol_names, nouts)
             # We execute the merge _after_ both branches, and pretend for a moment, that it
             # can see those values produced in child scopes.
             with _cond_merge(split_predicate):
-                for new_body_val, new_orelse_val in zip(body_outputs, orelse_outputs):
+                err_msg = ("Divergent data found in different branches of `if/else` control flow"
+                           " statement. Variables in all code paths are merged into common output"
+                           " batches. The values assigned to a given variable need to have the same"
+                           " nesting structure in every code path (both `if` branches).\n"
+                           "For example, if we define a variable as a tuple in one branch, it must"
+                           " be defined as a tuple of the same length in the other branch - the"
+                           " contents of the tuples may be different. If we define a variable as"
+                           " a dictionary, the other branch must define it as a dictionary with the"
+                           " same set of keys, the values may be different.\n")
+
+                try:
+                    tree.assert_same_structure(body_outputs, orelse_outputs, check_types=True)
+                except ValueError as e:
+                    # Suppress the original exception, add DALI explanation at the beginning,
+                    # raise the full error message.
+                    raise ValueError(err_msg + str(e)) from None
+                except TypeError as e:
+                    raise TypeError(err_msg + str(e)) from None
+
+                def merge_branches(new_body_val, new_orelse_val):
                     logging.log(9, (f"{this_condition_stack()._indent()}[IF] Inserting merge"
                                     f" at {this_condition_stack().stack_depth() -1}:"
                                     f" merge({new_body_val}, {new_orelse_val}, predicate="
                                     f"{split_predicate}."))
-                    merged = fn._conditional.merge(new_body_val, new_orelse_val,
-                                                   predicate=split_predicate)
-                    output_values.append(merged)
+                    return fn._conditional.merge(new_body_val, new_orelse_val,
+                                                 predicate=split_predicate)
+
+                output_values = tree.map_structure(merge_branches, body_outputs, orelse_outputs)
```

Comment:
1. Do we need `output_values = []` at L:515 then? 
2. The `body_outputs` and `orelse_outputs` are always some sort of tuple, right? Just sometimes with elements that may themselves be structures, but we'll never get here dictionary or plan DataNode? I am asking in the context of `output_values += init_state[nouts:]` in L:558

Reply:
Yes, we can probably remove the initialization, as we just overwrite it with something that is created as a list.

Yes, the state is a tuple/list that is assigned to list of state that are symbolic variables.